### PR TITLE
fix mirage

### PR DESCRIPTION
**Pull request by koshuang**
## Summray

* improve developer experience
  * vscode
    * auto format code on save
  * debugging
    * by adding `DEBUG` and `DEBUG_COLORS` environment to have better debugging logs. 
* fix bugs
* new features
  * add `getDebugger` on package `@onr/common`
  * move `mirage` to package `@onr/core`
* Breaking changes
  * 因為 core module 與 onr-core package 在使用時都是 `@onr/core` ，所以在 runtime 時只會拿到 module 裡的 core。暫時先透過調整 webpack 的設定，把 onr-core package 改成 `@core` 來使用。未來等全部都移植完畢之後，再去把他改回來。
  * 因為新的 next.js 的環境變數會使mirage在前端拿不到 `API_URL` ，導致 mirage 壞掉，解法就是加上 prefix `NEXT_PUBLIC_`
    * `API_URL` to `NEXT_PUBLIC_API_URL`
    * `API_KEY` to `NEXT_PUBLIC_API_KEY`

## Screenshots
Debug logs on server side and client side (only firefox supported)

Screenshot from 2021-05-25 11-28-45
Screenshot from 2021-05-25 11-27-50

## References
* https://github.com/visionmedia/debug#readme
* https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser

### CODE REVIEW

File: packages/next-starter/.env.example [hunk 1]
```diff
@@ -1,5 +1,7 @@
 NODE_ENV=development
 PROXY_MODE=local
+# DEBUG=onr:core:*
+# DEBUG_COLORS=true
```
Comment (l443018) [inline packages/next-starter/.env.example:4]: DEBUG=onr:core:* 效果是什麼呢？

File: packages/next-starter/README.md [hunk 2]
```diff
@@ -120,6 +120,18 @@ Then open http://localhost:3000/ to see your app.
...
+By setting environment variable `DEBUG`, you can see a specific logs of some packages on backend. For example,
+
+```
+DEBUG=onr:core:*
+```
+
+You will see all logs for `@onr/core` package.
+
+To see browser debug logs, you can use Firefox and then set `localPackage.debug = 'onr:core:*'.
+
```
Comment (l443018) [inline packages/next-starter/README.md:134]: localPackage 是在哪邊設定的呢？

File: packages/next-starter/.env.example [hunk 3]
```diff
@@ -1,11 +1,13 @@
 NODE_ENV=development
 PROXY_MODE=local
-API_URL=http://localhost:3000
-API_KEY=your_api_key
+# DEBUG=onr:core:*
+# DEBUG_COLORS=true
+NEXT_PUBLIC_API_URL=http://localhost:3000
+NEXT_PUBLIC_API_KEY=your_api_key
```
Comment (l443018) [inline packages/next-starter/.env.example:6]: 為何NEXT_PUBLIC_API_URL就可以拿到了呢？跟API_URL差異是？

File: packages/next-starter/README.md (hunk 2 above)
Comment (koshuang) [reply packages/next-starter/README.md:134]: 在 console

File: packages/next-starter/.env.example (hunk 3 above)
Comment (koshuang) [reply packages/next-starter/.env.example:6]: 在 Reference Link 上有， https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser

**Comment by coomysky** [comment]
r+ thanks!

File: packages/next-starter/.vscode/settings.json [hunk 4]
```diff
@@ -2,5 +2,30 @@
...
+  },
+  "[typescriptreact]": {
+    "editor.formatOnSave": true
+  },
+  "files.associations": {
+    "*.json": "jsonc"
+  },
+  "cSpell.words": [
+    "antd"
+  ]
```
Comment (rt-cff) [inline packages/next-starter/.vscode/settings.json:30]: Did we list all the vs code extensions used somewhere in doc?

File: packages/next-starter/src/modules/shared/mirage/mirage.ts [hunk 5]
```diff
@@ -75,6 +75,14 @@ export function makeServer({ environment = 'test' } = {}) {
...
         registerRoute(RoutesRegistration[route]);
       }
+
+       this.passthrough((request) => {
+        if (
+          request.url === "/_next/static/development/_devPagesManifest.json"
+        ) {
+          return true;
+        }
+      });
```
Comment (rt-cff) [inline packages/next-starter/src/modules/shared/mirage/mirage.ts]: How about declaring a const variable at the top of the file, having all the whitelisted passthrough path? 

Actually, I believe, bypassing all path with `/_next`, should be there by default?

File: packages/next-starter/src/pages/_app.tsx [hunk 6]
```diff
@@ -1,6 +1,5 @@
-import { makeServer } from '@onr/shared/mirage';
-
 if (process.env.NODE_ENV === 'development' && process.env.NEXT_PUBLIC_ENABLE_MIRAGE === 'true') {
+  const { makeServer } = require('@onr/shared/mirage');
   makeServer({ environment: 'development' });
 }
```
Comment (rt-cff) [inline packages/next-starter/src/pages/_app.tsx]: I think we can keep the import syntax? although that would make the code looks a bit complex
```
(async() => {
   const {makeServer} = await import('@onr/shared/mirage');
   ....
})()
```

**Review by rt-cff — COMMENTED** [review]
r+

File: packages/next-starter/src/pages/_app.tsx (hunk 6 above)
Comment (rt-cff) [reply packages/next-starter/src/pages/_app.tsx]: or use Promise.then syntax

File: packages/next-starter/.vscode/settings.json (hunk 4 above)
Comment (koshuang) [reply packages/next-starter/.vscode/settings.json:30]: Good idea!

File: packages/onr-core/src/mock/mirage.ts [hunk 7]
```diff
@@ -1,22 +1,26 @@
 import { Server } from 'miragejs';
-import MOCK_JSON, { ModelsDeclaration, RoutesRegistration } from '../__mock__';
+import { getDebugger } from '@onr/common';
 
-export function makeServer({ environment = 'test' } = {}) {
+const debug = getDebugger('onr:core:mirage');
+
+export function makeServer({ environment = 'test', seeds = {}, models = {}, routes = {} } = {}) {
```
Comment (TianyiLi) [inline packages/onr-core/src/mock/mirage.ts:6]: should have more clarity type hint
for `seeds`, seems it limited the input object should have `s` in the end, how about using [Template literal type](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) to solve the problem?
and also for `models` and `routes` should be declared as `Record<string, unknown>`.

Comment (TianyiLi) [reply packages/onr-core/src/mock/mirage.ts:6]: And also, I'm not like `${string}s`  this pattern, it could be a bug when developing the mirage schema

**Review by TianyiLi — COMMENTED** [review]
r+

**Pull request merged**